pkg/controllers/provisioner/v1alpha1/allocation/allocator.go

```diff
@@ -0,0 +1,69 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package allocation
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/awslabs/karpenter/pkg/cloudprovider"
+	v1 "k8s.io/api/core/v1"
+)
+
+var _ Allocator = &GreedyAllocator{}
+
+// GreedyAllocator iteratively assigns pods to scheduling groups and then creates capacity for each group.
+type GreedyAllocator struct {
+	Capacity cloudprovider.Capacity
+}
+
+//
+func (a *GreedyAllocator) Allocate(pods []*v1.Pod) error {
+	// 1. Separate pods into scheduling groups
+	groups := a.getSchedulingGroups(pods)
```

Comment:
How do we make sure we are not provisioning multiple times for the same pod?
Reconciler will check every 10-15 seconds for pending pods and we might see the same pod again.

Reply:
You're totally right -- we need a solution for that. I see that as an implementation detail of whatever passes pods into this function.